File: frontend/src/types/qa.ts
export interface Source {
  chunk_id: string;
  document_id: string;
  original_filename: string;
  file_type: string;
  page: number | null;
  chunk_index: number;
  text: string;
  text_preview: string;
  distance: number;
  score: number;
  retrieval_source: "vector" | "keyword" | "hybrid" | string;
  final_score: number;
  vector_rank: number | null;
  keyword_rank: number | null;
  vector_score: number;
  keyword_score: number;
}

export interface AskRequest {
  question: string;
  top_k?: number;
  document_id?: string | null;
  document_ids?: string[] | null;
  knowledge_base_ids?: string[] | null;
  conversation_id?: string | null;
  max_distance?: number | null;
  hybrid_search_enabled?: boolean;
  vector_candidate_count?: number | null;
  keyword_candidate_count?: number | null;
}

export interface AskResponse {
  question: string;
  answer: string;
  sources: Source[];
  selected_document_ids: string[];
}

export type ChatRole = "user" | "assistant" | "error";

export interface RetrievalSettings {
  topK: number;
  maxDistance: number | null;
}

export interface AssistantRequestContext {
  question: string;
  selectedDocumentIds: string[];
  knowledgeBaseIds: string[];
  conversationId: string | null;
  retrievalSettings: RetrievalSettings;
}

export interface ChatMessage {
  id: string;
  role: ChatRole;
  content: string;
  created_at?: string;
  sources?: Source[];
  selected_document_ids?: string[];
  knowledge_base_ids?: string[];
  conversation_id?: string | null;
  top_k?: number | null;
  max_distance?: number | null;
  requestContext?: AssistantRequestContext;
}

export interface ChatHistoryResponse {
  messages: ChatMessage[];
}

File: backend/app/schemas/qa.py
from pydantic import BaseModel, Field


class SearchRequest(BaseModel):
    question: str = Field(min_length=1)
    top_k: int = Field(default=5, ge=1, le=20)
    document_id: str | None = None
    document_ids: list[str] | None = None
    knowledge_base_ids: list[str] | None = None
    conversation_id: str | None = None
    max_distance: float | None = Field(default=None, ge=0, le=2)
    hybrid_search_enabled: bool = True
    vector_candidate_count: int | None = Field(default=None, ge=1, le=100)
    keyword_candidate_count: int | None = Field(default=None, ge=1, le=100)


class AskRequest(BaseModel):
    question: str = Field(min_length=1)
    top_k: int = Field(default=5, ge=1, le=20)
    document_id: str | None = None
    document_ids: list[str] | None = None
    knowledge_base_ids: list[str] | None = None
    conversation_id: str | None = None
    max_distance: float | None = Field(default=None, ge=0, le=2)
    hybrid_search_enabled: bool = True
    vector_candidate_count: int | None = Field(default=None, ge=1, le=100)
    keyword_candidate_count: int | None = Field(default=None, ge=1, le=100)


class SearchResult(BaseModel):
    chunk_id: str
    document_id: str
    original_filename: str = ""
    file_type: str = ""
    page: int | None = None
    chunk_index: int
    text: str
    text_preview: str = ""
    distance: float
    score: float = 0.0
    retrieval_source: str = "vector"
    final_score: float = 0.0
    vector_rank: int | None = None
    keyword_rank: int | None = None
    vector_score: float = 0.0
    keyword_score: float = 0.0


class SearchResponse(BaseModel):
    question: str
    results: list[SearchResult]
    selected_document_ids: list[str] = []


class AskResponse(BaseModel):
    question: str
    answer: str
    sources: list[SearchResult]
    selected_document_ids: list[str] = []
